pkg/controller/jobs/pod/pod_controller.go

```diff
@@ -147,91 +203,445 @@ func (p *Pod) PodSets() []kueue.PodSet {
 	}
-	if err == nil || apierrors.IsNotFound(err) {
-		return true, nil
+
+	for i := range podsInGroup {
+		if podsInGroup[i].ObjectMeta.DeletionTimestamp.IsZero() {
+			return false
+		}
 	}
-	return false, err
+
+	return true
 }
 
 func SetupIndexes(ctx context.Context, indexer client.FieldIndexer) error {
 	return jobframework.SetupWorkloadOwnerIndex(ctx, indexer, gvk)
 }
 
 func (p *Pod) Finalize(ctx context.Context, c client.Client) error {
-	if controllerutil.RemoveFinalizer(p.Object(), PodFinalizer) {
-		if err := c.Update(ctx, p.Object()); err != nil {
+	groupName := p.groupName()
+
+	var podsInGroup corev1.PodList
+	if groupName == "" {
+		podsInGroup.Items = append(podsInGroup.Items, *p.Object().(*corev1.Pod))
+	} else {
+		if err := c.List(ctx, &podsInGroup, client.MatchingLabels{
+			GroupNameLabel: groupName,
+		}, client.InNamespace(p.pod.Namespace)); err != nil {
 			return err
 		}
 	}
 
+	for _, pod := range podsInGroup.Items {
+		if controllerutil.RemoveFinalizer(&pod, PodFinalizer) {
+			if err := c.Update(ctx, &pod); err != nil {
+				return err
+			}
+		}
+	}
+
 	return nil
 }
 
 func (p *Pod) Skip() bool {
 	// Skip pod reconciliation, if managed label is not set
-	if v, ok := p.GetLabels()[ManagedLabelKey]; !ok || v != ManagedLabelValue {
+	if v, ok := p.pod.GetLabels()[ManagedLabelKey]; !ok || v != ManagedLabelValue {
 		return true
 	}
-
 	return false
 }
 
+func (p *Pod) groupName() string {
+	return p.Object().GetLabels()[GroupNameLabel]
+}
+
+func (p *Pod) groupTotalCount() (int, error) {
+	if p.groupName() == "" {
+		return 0, fmt.Errorf("pod doesn't have a '%s' label", GroupNameLabel)
+	}
+
+	gtcAnnotation, ok := p.Object().GetAnnotations()[GroupTotalCountAnnotation]
+	if !ok {
+		return 0, fmt.Errorf("failed to extract '%s' annotation",
+			GroupTotalCountAnnotation)
+	}
+
+	gtc, err := strconv.Atoi(gtcAnnotation)
+	if err != nil {
+		return 0, err
+	}
+
+	if gtc < 1 {
+		return 0, fmt.Errorf("incorrect annotation value '%s=%s': group total count should be greater than zero",
+			GroupTotalCountAnnotation, gtcAnnotation)
+	}
+
+	return gtc, nil
+}
+
+// Load loads all pods in the group
+func (p *Pod) Load(ctx context.Context, c client.Client, key types.NamespacedName) (removeFinalizers bool, err error) {
+	if err := c.Get(ctx, key, &p.pod); err != nil {
+		return apierrors.IsNotFound(err), err
+	}
+	groupName := p.groupName()
+	p.isGroup = groupName != ""
+	if !p.isGroup {
+		return !p.pod.DeletionTimestamp.IsZero(), nil
+	}
+
+	if err := c.List(ctx, &p.list, client.MatchingLabels{
+		GroupNameLabel: groupName,
+	}, client.InNamespace(key.Namespace)); err != nil {
+		return false, err
+	}
+
+	wl, _, err := p.FindMatchingWorkloads(ctx, c)
+	if err != nil {
+		return false, err
+	}
+
+	if wl != nil {
+		// If workload is finished remove its finalizer
+		if apimeta.IsStatusConditionTrue(wl.Status.Conditions, kueue.WorkloadFinished) {
+			return true, nil
+		}
+	} else {
+		if p.isStopped() {
```

Comment:
We can't remove the Workload or Pod finalizers even if all the pods have deletion timestamp.

Otherwise the Workload would be removed and it wouldn't be possible to send replacement pods.

Reply:
The finalizers are removed here if workload is nil and all the pods have deletion timestamp. There's nothing to replace if workload has been already removed. 

But this code needs a change anyway. `FindMatchingWorkloads` could return some workloads to delete, in this case we shouldn't finalize the pods.

Reply:
Ah gotcha. However, it looks like we are duplicating code. We already finalize the job in the following scenarios:

- If the workload is finished: https://github.com/kubernetes-sigs/kueue/blob/9692f3ac5ccbcfc10c30a9e9703249d7549babac/pkg/controller/jobframework/reconciler.go#L263
- If there is no matching Workload: https://github.com/kubernetes-sigs/kueue/blob/9692f3ac5ccbcfc10c30a9e9703249d7549babac/pkg/controller/jobframework/reconciler.go#L454

So it looks like this logic is unnecessary here

Reply:
This code is there for the case, when wl is deleted. 

First reconcile will hit this branch, stop the job, finalize the wl and return:
https://github.com/kubernetes-sigs/kueue/blob/9692f3ac5ccbcfc10c30a9e9703249d7549babac/pkg/controller/jobframework/reconciler.go#L247

Second reconcile will finalize the pods after `ComposableJob.Load` call.

We could finalize the pods on the first reconcile, but I don't think it's right from the interface perspective. Stop != Delete for the generic job. It's only true for pods. That's why the logic to finalize the pods if wl is nil is in `ComposableJob.Load` and not in the generic reconciler.

Reply:
> We could finalize the pods on the first reconcile.

Please let me know what do you think about this idea.

Reply:
I see what you mean, for the case when the workload is deleted. However, the case when the Workload is finished is already covered.

Then, we need a single line:

```
return wl != nil && p.isStopped(), nil
```

Reply:
And accompany it with a comment about why we need to finalize in that case.